nova/virt/vmwareapi/vmops.py

```diff
@@ -2268,46 +2443,166 @@ def _relocate_vm(self, vm_ref, context, instance, network_info,
+        allowed_ds_types = ds_util.get_allowed_datastore_types(
+            image_meta.properties.hw_disk_type)
+
+        # Determine whether we should connect to the dest vCenter endpoint
+        # (if the instance is to be migrated outside of current vCenter)
+        remote = self._is_remote(dest)
+
+        service_spec = None
+        if remote:
+            session = self._create_dest_session(dest)
+            # set the instance_uuid needed in _get_service_locator_spec
+            dest.update({
+                'instance_uuid':
+                    session.vim.service_content.about.instance_uuid})
+            service_spec = self._get_service_locator_spec(dest)
+            cluster = vutil.get_moref(dest['cluster'],
+                                      'ClusterComputeResource')
+            datastore_regex = dest['datastore_regex']
+            # For migration to other vCenter service the disk_move_type
+            # needs to be moveAllDiskBackingsAndDisallowSharing
+            disk_move_type = "moveAllDiskBackingsAndDisallowSharing"
+        else:
+            session = self._session
+            cluster = self._cluster
+            datastore_regex = self._datastore_regex
+            disk_move_type = "moveAllDiskBackingsAndAllowSharing"
+
+        res_pool = vm_util.get_res_pool_ref(session, cluster)
+        datastore = ds_util.get_datastore(session, cluster,
+                                          datastore_regex,
+                                          storage_policy,
+                                          allowed_ds_types)
+        dc_info = self.get_datacenter_ref_and_name(datastore.ref)
+        folder = self._get_project_folder(dc_info, instance.project_id,
+                                          'Instances', session=session)
+
+        rel_spec = vm_util.relocate_vm_spec(self._session.vim.client.factory,
+                                            disk_move_type=disk_move_type,
+                                            res_pool=res_pool, folder=folder,
+                                            datastore=datastore.ref,
+                                            service=service_spec)
+        vm_name = self._get_migration_vm_name(instance)
+        cloned_vm = self._clone_vm(vm_ref, rel_spec, folder, name=vm_name)
+        LOG.info("Cloned VM with temporary name %s", vm_name,
+                 instance=instance)
+        try:
+            # keep track of the old vm by assigning migration.uuid to it
+            vm_util.change_vm_instance_uuid(self._session, vm_ref,
+                                            migration.uuid)
+            vm_util.change_vm_instance_uuid(session, cloned_vm, instance.uuid)
+        except Exception:
+            with excutils.save_and_reraise_exception():
+                vm_util.destroy_vm(session, instance, cloned_vm)
+        vm_name = "%s_source" % vm_util._get_vm_name(instance.display_name,
+                                                     instance.uuid)
```

Comment:
Better use only `f"{instance.uuid}_source"` here, so we don't get over a possible limit in the name - because `display_name` might be very long (for k8s) and we shorten it already in `_get_vm_name()`.